cmd/state-svc/internal/resolver/analytics/analytics.go

```diff
@@ -225,36 +228,64 @@ func (r *Resolver) eventLoop() {
 	tick := time.NewTicker(time.Minute * 5)
 	defer tick.Stop()
 
+	projectIDMap := make(map[string]string)
+
 	// flush the deferred data initially
-	if err := r.flushDeferred(); err != nil {
+	if err := r.flushDeferred(projectIDMap); err != nil {
```

Comment:
This would give these events the project ID at the time that we flush the events. It should be set at the time that the event was triggered. Otherwise you might end up with the wrong project.

Reply:
If we set the project id _before_ writing them to the "deferred" file, we'll have to resolve the project id in the client (State Tool) process though.  That is we would have to send the API request with every new State Tool invocation.  That might not be the biggest issue if we indeed get rid of "forced" deferred events though:  Then, we'll only do that in occasions where the State Tool does not have a connection to the `state-svc` yet which should almost never happen.  But by adding such an analytics call, we would involuntarily pull in an API call then...

Reply:
In the latest iteration, I am setting the projectID [as early as possible in the state-svc](https://github.com/ActiveState/cli/blob/603be56e975c4964104344d7bc3053f8f324b1cb/cmd/state-svc/internal/resolver/analytics/analytics.go#L133), but not in the `DefaultClient` yet for the reasons explained above.

Note, however, that this takes the `projectID` resolution out of the event loop, such that I had to [introduce a mutex](https://github.com/ActiveState/cli/blob/603be56e975c4964104344d7bc3053f8f324b1cb/cmd/state-svc/internal/resolver/analytics/analytics.go#L252-L254)